src/integration/aggregator/aggregator.go

```diff
@@ -0,0 +1,302 @@
+	TestAggregatorDBNodeConfig = `
+db: {}
+`
+
+	// TestAggregatorCoordinatorConfig is the test config for the coordinator.
+	TestAggregatorCoordinatorConfig = `
+listenAddress: 0.0.0.0:7202
+carbon:
+  ingester:
+    listenAddress: "0.0.0.0:7204"
+    rules:
+      - pattern: .*
+        aggregation:
+          type: mean
+        policies:
+          - resolution: 10s
+            retention: 6h
+clusters:
+  - namespaces:
+      - namespace: default
+        type: unaggregated
+        retention: 48h
+      - namespace: aggregated
+        type: aggregated
+        resolution: 10s
+        retention: 6h
+downsample:
+  rules:
+    rollupRules:
+      - name: "requests per second by status code"
+        filter: "__name__:http_requests app:* status_code:* endpoint:*"
+        transforms:
+          - transform:
+              type: "PerSecond"
+          - rollup:
+              metricName: "http_requests_by_status_code"
+              groupBy: ["app", "status_code", "endpoint"]
+              aggregations: ["Sum"]
+        storagePolicies:
+          - resolution: 10s
+            retention: 6h
+storeMetricsType: true
+`
+
+	// TestAggregatorAggregatorConfig is the test config for the aggregators.
+	TestAggregatorAggregatorConfig = `
+`
+
+	// defaultCarbonPort is the default port of coordinator to receive carbon metrics.
+	defaultCarbonPort = 7204
+)
+
+var (
+	errEmptyResult = errors.New("empty query result")
+	errQueryResult = errors.New("wrong query result")
+)
+
+// TestAggregatedGraphiteMetric tests the write and read of aggregated graphtie metrics.
+func TestAggregatedGraphiteMetric(t *testing.T, m3 resources.M3Resources) {
+	var (
+		carbonName         = "foo.bar.baz"
+		carbonTarget       = "foo.bar.*"
+		carbonLow          = float64(40)
+		carbonHigh         = float64(44)
+		expectedCarbonMean = float64(42)
+	)
+
+	doneCh := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-doneCh:
+				return
+			default:
+				require.NoError(t, m3.Coordinator().WriteCarbon(defaultCarbonPort, carbonName, carbonLow, time.Now()))
+				require.NoError(t, m3.Coordinator().WriteCarbon(defaultCarbonPort, carbonName, carbonHigh, time.Now()))
+				time.Sleep(1 * time.Second)
+			}
+		}
+	}()
+
+	require.NoError(t, resources.Retry(func() error {
+		return verifyGraphiteQuery(m3, doneCh, carbonTarget, expectedCarbonMean)
+	}))
+}
+
+func verifyGraphiteQuery(m3 resources.M3Resources, doneCh chan struct{}, target string, expected float64) error {
+	datapoints, err := m3.Coordinator().GraphiteQuery(resources.GraphiteQueryRequest{
+		Target: target,
+		From:   time.Now().Add(-1000 * time.Second),
+		Until:  time.Now(),
+	})
+	if err != nil {
+		return err
+	}
+	nonNullDPs := filterNull(datapoints)
+	if len(nonNullDPs) == 0 {
+		return errEmptyResult
+	}
+	if v := *nonNullDPs[0].Value; v != expected {
+		return fmt.Errorf("wrong datapoint result: expected=%f, actual=%f", expected, v)
+	}
+	doneCh <- struct{}{}
```

Comment:
You'll want to call this with `defer` so that the goroutine doing the writes gets shutdown even upon error.